lib: posix: add include/posix to search path and drop posix header prefix

cfriedt opened the pull request:
It is a pain point to prefix standard POSIX APIs (e.g. `#include <zephyr/posix/unistd.h>`). This is true, in particular, when integrating third-party code into an external module. However, using the POSIX API should also not be a default in Zephyr.

With this, if users choose to include standard POSIX headers via Kconfig, then those headers will be part of the default search path. The headers themselves do not move. POSIX headers will not appear in the default search path unless so configured.
    
Additional changes:
* removed `posix/time.h` to mitigate need for `#include_next` workaround
* gathered `NSEC_PER_MSEC` and `USEC_PER_MSEC` defs into`sys_clock.h`
* restructured some syscalls (unistd, ioctl, clock)
* sprinkled `#include <unistd.h>` where `close()` is used or use `zsock_close()` instead
* removed a fair amount of duplicate code
* additional improvements planned to improve code sharing / coexistence with net API
* suggested minor changes to picolibc https://github.com/zephyrproject-rtos/zephyr/issues/46910

Fixes #43998

cfriedt commented:
- [ ] civetweb module: change `posix/time.h` to `time.h`

rlubos commented on include/net/socket.h:
```diff
@@ -672,11 +672,6 @@ static inline int socketpair(int family, int type, int proto, int sv[2])
 	return zsock_socketpair(family, type, proto, sv);
 }
 
-static inline int close(int sock)
```
It's expected now to include POSIX header even if there's no intention to use `CONFIG_POSIX_API`?

cfriedt replied:
Selected by `CONFIG_NET_SOCKETS_POSIX_NAMES` as well, but yeah.

Could always use `zsock_close()` when we really want to make it clear we're not using POSIX.

Aside from that: looks like POSIX, smells like POSIX, .. :shrug:

cfriedt replied:
@rlubos - done!

cfriedt replied:
@rlubos @pfalcon 
I almost wonder if it's time to simply remove
`CONFIG_NET_SOCKETS_POSIX_NAMES`.. 🤔 Is there a compelling reason to keep it?

If devs want to use POSIX API calls as a shorthand, then it makes sense to select `CONFIG_POSIX_API`. For networking calls, they really are just simple wrappers around the `zsock_*` versions. For other file-related calls, they're just simple wrappers around the related fdtable calls.

Otherwise use the native Zephyr variants.

cfriedt replied:
> Is it expected now to include POSIX header even if there's no intention to use `CONFIG_POSIX_API`?

Technically it doesn't have to be (e.g. `#include <posix/unistd.h>` will still work), but if the intent is to call a standard POSIX function, then it makes sense to include a standard POSIX header and to cater to portability rather than 3rd-party APIs that knowingly conflict with standards.

cfriedt commented:
> Yeah, there must have been a reason why it was done that way... And certainly, nobody will test how it will work with various 3rd-party projects... Good plan anyway.

Technically, 3rd-party projects are responsibility of 3rd parties. Best practice is to pin to a tagged release, so if anyone is pinning to `HEAD` of `main`, changes *do* happen and the build does break on occasion. :shrug: 

I would much rather pay off tech-debt today than after another 4 years when the interest rate has gone up :shrug:.

pfalcon commented:
> @rlubos @pfalcon 
> I almost wonder if it's time to simply remove
> `CONFIG_NET_SOCKETS_POSIX_NAMES`.. 🤔 Is there a compelling reason to
> keep it?

I don't know if there *is* a compelling reason to keep it, but I can
remind why it *was* done like that: there're known cases when vendors
supply for their "network processors" non-namespaced sockets-like API.
To allow to both integrate with such 3rd-party solutions and still use
native Zephyr networking stack (e.g. on different network interfaces),
Zephyr's implementation was made namespaced, with ability to provide
native POSIX aliases.

-- 
Best regards,
 Paul                          ***@***.***

cfriedt commented on include/net/socket.h:
```diff
@@ -957,6 +769,14 @@ struct net_socket_register {
 
+#ifdef CONFIG_NET_SOCKETS_POSIX_NAMES
+/* These prefixed with posix as they can be used without CONFIG_POSIX_API */
+#include <posix/arpa/inet.h>
```
@rlubos - I might remove `CONFIG_NET_SOCKET_POSIX_NAMES` from `lib/posix/CMakeLists.txt` and add 
```cpp
#include <posix/fcntl.h> // fcntl(2)
#include <posix/unistd.h> // close(2)
// also maybe remove <posix/sys/socket.h> here, since there is already <net/socket_select.h>
```

What are your thoughts on that?

cfriedt replied:
Technically, a lot of the POSIX stuff is just defined directly in `net/socket.h`, but some things should probably be prefixed (even though there is a 1:1 relationship). E.g. `TCP_NODELAY`.

cfriedt commented:
fixed merge conflict / rebased after #44320

cfriedt commented:
* default to `CONFIG_POSIX_MAX_FDS=7`

cfriedt commented:
Hey - I can use socketpair on `native_posix_64` now!

cfriedt commented:
* added `np_get_libc()` function call available with `CONFIG_ARCH_POSIX` which allows us to refer to native libc symbols (functions) that have been overridden by minimal libc symbols.

aescolar reviewed (commented):
@cfriedt This PR is mutating a lot and I don't think I follow were you are trying to get it :)

Anyhow, I'm not incredibly fond of that dynamic loading of the host libc for the posix architecture with the need to define trampolines by default.
Have you seen the direction proposed here: https://github.com/zephyrproject-rtos/zephyr/issues/24685  ?

cfriedt commented:
> @cfriedt This PR is mutating a lot and I don't think I follow were you are trying to get it :)

@aescolar the point of this PR is to drop the `posix/` prefix from standard includes.

The PR has mutated over time because the codebase has mutated over time, I'm really just trying to keep up with `main`. 

Additionally, other modules have needed to be modified to account for the removing the `posix/` prefix.

> Anyhow, I'm not incredibly fond of that dynamic loading of the host libc for the posix architecture with the need to define trampolines by default. Have you seen the direction proposed here: #24685

Currently there is only 1 trampoline needed for all of the tests that are run in CI (for `fcntl()` specifically, in the initialization for `uart_native_posix.c`). It's a portable solution that only relies on the standard POSIX API.

Given that Zephyr overrides many standard symbols via the minimal libc (which is actually what we want), then it makes a lot of sense as well.

What makes you not incredibly fond of it?

Looking at #24685 now. I'll drop some comments there as well.

aescolar commented:
@cfriedt (Just as info as you may not know it: I originally wrote the posix arch and native_posix and I'm still their maintainer).
The reason why I referred to https://github.com/zephyrproject-rtos/zephyr/issues/24685 is because you are facing a fundamental limitation in the way the posix arch was built (that when plainly linking together the posix shim API with the posix arch, and/or one of the provided embedded libC's and the host libC, there is symbol conflicts). This was understood since the very beginning. The posix_cheats header was just a very simple way of trying to avoid it, but it caused trouble when continuing to expand the POSIX API shim and running it on the POSIX arch, and that is why it was disabled instead of digging deeper into that hole.
The proposal in that referred issue provides what (at least to me) seems like a nice and clean solution. I just never got to implement it due to lack of priority.
On the other hand the way you are trying to work around the issue with the dynamic loading of the host libC, is not addressing the issue in itself but partially hiding it, and I do not think will not work well in the long run. In any case the issues that lead to disabling the combination of POSIX API shim on the POSIX arch are not addressed here (It has always been possible to work around each issue with some hacks).
In any case, it is of course up to you if you want to understand and consider what is proposed there, or you prefer to continue down this path. But please note that with the direction proposed in this PR today I would need to reject it.

cfriedt commented:
> @cfriedt (Just as info as you may not know it: I originally wrote the posix arch and native_posix and I'm still their maintainer). 

@aescolar - I'm well aware.

> The reason why I referred to #24685 is because you are facing a fundamental limitation in the way the posix arch was built (that when plainly linking together the posix shim API with the posix arch, and/or one of the provided embedded libC's and the host libC, there is symbol conflicts). This was understood since the very beginning. The posix_cheats header was just a very simple way of trying to avoid it, but it caused trouble when continuing to expand the POSIX API shim and running it on the POSIX arch, and that is why it was disabled instead of digging deeper into that hole.

It's not really that much of a limitation.

The synopsis is that we want to abstract away the host OS system calls (and a large part of the system C library) so that we can exercise Zephyr code paths rather than relying on host code paths.

The `native_posix` architecture should perform that abstraction in a way that is completely invisible to applications, and I think we would both agree that it could be improved.

That topic is actually quite tangential to point of this PR, and I believe the only reason you're commenting here, and maybe attacking the person versus critiquing code is because of the use of `dlopen()`, so I'm going file a bug and open another PR for that purpose. I'll mark this PR as depending on that issue being resolved, and I'll respond to your comments there.

cfriedt committed:
lib: posix: add include/posix to search path and drop posix header prefix

It is a pain point to prefix standard POSIX APIs (e.g.
`#include <posix/unistd.h>`). This is true, in particular,
when integrating third-party code into an external module.

With this, if users choose to include standard POSIX headers via Kconfig,
then those headers will be part of the default search path. The headers
themselves do not move. POSIX headers will not appear in the default
search path unless so configured.

Additional changes:
* removed `posix/time.h` to circumvent `#include_next` workaround
* gathered `NSEC_PER_MSEC` and `USEC_PER_MSEC` defs into `sys_clock.h`
* restructured some syscalls (unistd, ioctl, clock_getttime)
* sprinkled `#include <unistd.h>` where `close()` is used
* removed a fair amount of duplicate code

Fixes #43998

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
Will require some rework due to picolibc merge

jukkar commented:
> > > CONFIG_NET_SOCKETS_POSIX_NAMES
> > > I don't know if there _is_ a compelling reason to keep it, but I can remind why it _was_ done like that: there're known cases when vendors supply for their "network processors" non-namespaced sockets-like API. To allow to both integrate with such 3rd-party solutions and still use native Zephyr networking stack (e.g. on different network interfaces), Zephyr's implementation was made namespaced, with ability to provide native POSIX aliases.
> 
> @pfalcon - do you know of a specific instance of that scenario that is supported? E.g. a particular vendor or sample application?

Not Paul, but at least Atmel ASF HAL is "buggy" as it does not namespace its APIs properly for winc1500 driver.
See `modules/hal/atmel/asf/common/components/wifi/winc1500/socket/include/socket.h` for details.

keith-packard commented:
What's the end goal here? Do you want Zephyr to use the native POSIX syscall API as defined by the C library headers? Or do you want Zephyr to use inline functions in a Zephyr-private header? If picolibc is built as part of the toolchain (as required for complete C++ support), then picolibc will be using a small number of POSIX syscalls with regular function linkage, so at least those functions would need to be provided with regular linkage.

Hrm. Maybe what you want is the GCC extern inline semantics. In that world, functions are provided as *both* static inline versions and functions with regular linkage. Code using the C library headers would use the non-inline functions, while code using Zephyr headers would get the static inline definitions and use them instead. That should limit use of the extern functions to just those calls from picolibc itself.

cfriedt commented:
> What's the end goal here? Do you want Zephyr to use the native POSIX syscall API as defined by the C library headers? Or

Currently, there are no plans of this touching `native_posix`, so `CONFIG_POSIX_API` and `CONFIG_ARCH_POSIX` are still mutually exclusive (although I have outlined some work in #45100).

> I think it would be nice if Zephyr code was just using the standard POSIX calls directly, not calls prefixed with some zephyr namespace. That would indicate that the POSIX semantics were in play to even a casual developer.

The main goals of `native_posix` are [here](https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/index.html#rationale-for-this-port), so I would think that we would ultimately prefer to continue using it to test / debug / instrument Zephyr's implementation of things, including the POSIX API, libc, etc.

The part of the arch / api problem is the global namespace, of course, so native libc symbols conflict with the Zephyr implementations (or vice-versa, depending on your PoV).

> Hrm. Maybe what you want is `gnu_inline` semantics. In that world, functions are provided as _both_ static inline versions 

Personally, I would like to implement that using completely plain, POSIX / ISO C. No ELF-isms, gnu-isms, gcc-isms, etc. Mainly because I'm also working on the macOS port. Everything that isn't portable kind of sticks out like a sore thumb :P

keith-packard commented:
> This change mainly allows the application to opt-in to `#include <unistd.h>` rather than `#include <zephyr/posix/unistd.h>`, and it is Zephyr's header (not the host's).

That's not valid -- the C library will be built with a different version of unistd.h and that may have conflicting definitions of various values, leading to failures when applications call the C library and it then uses symbols defined differently in its own unistd.h

cfriedt commented:
> > @keith-packard - can you be a bit more specific?
> The C library uses POSIX functions for various operations, in particular, picolibc uses:

The hazards look entirely manageable.

**Probably should not be used in the Zephyr (yet?)**
isatty

**Probably should not be used in Zephyr**
getpid
kill

**If there is any discrepancy, we will likely use the picolibc definitions**
lseek
lseek64
fstat
open

**Likely fine (Note: we use libc's <time.h>):**
read
unlink
write
gettimeofday
close
_exit

keith-packard commented:
> More or less what I anticipated as well, because newlib had similar overlap. The hazards look entirely manageable though, @keith-packard. I would say the one that could potentially be (hand-wavy) is open, but mainly because of `fcntl.h`.

Zephyr *currently* has conflicting definitions for many of these APIs as it defines them as aliases for static inline functions. That means code including picolibc headers will not get the correct definitions, they'll get external function references, which may (or may not) work, but will certainly be less efficient.

I think what would be best would be for the standard posix headers to define the standard posix APIs generating the right code to operate in Zephyr. That means inline code for syscalls, for instance.

Given that the Zephyr SDK is already using a new GNU triplet including '-zephyr-', it's reasonable to expect the libraries in that SDK to be zephyr-specific, so if the C library is built as part of that toolchain, it can (and should) understand the Zephyr syscall interface, defining the functions correctly for use in Zephyr. Similarly, a C library built as part of Zephyr should use the Zephyr syscall interface.

Including hacks in Zephyr to work around libraries not designed for it is reasonable where you can't fix the library. In this case, the upstream projects would all be happy to accept Zephyr OS support.

cfriedt commented:
> > More or less what I anticipated as well, because newlib had similar overlap. The hazards look entirely manageable though, @keith-packard. I would say the one that could potentially be (hand-wavy) is open, but mainly because of `fcntl.h`.
> 
> Zephyr _currently_ has conflicting definitions for many of these APIs as it defines them as aliases for static inline functions.

This change fixes that. 

> I think what would be best would be for the standard posix headers to define the standard posix APIs generating the right code to operate in Zephyr. That means inline code for syscalls, for instance.

This change fixes that too.

> Given that the Zephyr SDK is already using a new GNU triplet including '-zephyr-', it's reasonable to expect the libraries in that SDK to be zephyr-specific, so if the C library is built as part of that toolchain, it can (and should) understand the Zephyr syscall interface, defining the functions correctly for use in Zephyr. Similarly, a C library built as part of Zephyr should use the Zephyr syscall interface.

That is a reasonable long-term goal. Some work depends on this being resolved in the short-term.

> Including hacks in Zephyr to work around libraries not designed for it is reasonable where you can't fix the library. In this case, the upstream projects would all be happy to accept Zephyr OS support.

Upstreamers welcome :-)

cfriedt commented:
Currently doing a rework of this to separate it into a series of several easily-reviewable commits.
